src/thunderbots/software/radio_communication/mrf/dongle.cpp

```diff
@@ -0,0 +1,631 @@
+    device.set_interface_alt_setting(radio_interface, normal_altsetting);
+
+    // Prepare the available message IDs for allocation.
+    for (unsigned int i = 0; i < 256; ++i)
+    {
+        free_message_ids.push(static_cast<uint8_t>(i));
+    }
+
+    // Submit the message delivery report transfers.
+    for (auto &i : mdr_transfers)
+    {
+        i.reset(new USB::BulkInTransfer(device, 1, 8, false, 0));
+        i->signal_done.connect(sigc::mem_fun(this, &MRFDongle::handle_mdrs));
+        i->submit();
+    }
+
+    // Submit the received message transfers.
+    for (auto &i : message_transfers)
+    {
+        i.reset(new USB::BulkInTransfer(device, 2, 105, false, 0));
+        i->signal_done.connect(sigc::bind(sigc::mem_fun(this, &MRFDongle::handle_message),
+                                          sigc::ref(*i.get())));
+        i->submit();
+    }
+
+    // Submit the estop transfer.
+    status_transfer.signal_done.connect(sigc::mem_fun(this, &MRFDongle::handle_status));
+    status_transfer.submit();
+}
+
+void MRFDongle::handle_libusb_events()
+{
+    context.handle_usb_events();
+}
+
+MRFDongle::~MRFDongle()
+{
+    // Mark USB device as shutting down to squelch cancelled transfer warnings.
+    device.mark_shutting_down();
+}
+
+void MRFDongle::beep(unsigned int length)
+{
+    pending_beep_length = std::max(length, pending_beep_length);
+    if (!beep_transfer && pending_beep_length)
+    {
+        beep_transfer.reset(new USB::ControlNoDataTransfer(
+            device, LIBUSB_REQUEST_TYPE_VENDOR | LIBUSB_RECIPIENT_INTERFACE,
+            MRF::CONTROL_REQUEST_BEEP, static_cast<uint16_t>(pending_beep_length),
+            static_cast<uint16_t>(radio_interface), 0));
+        beep_transfer->signal_done.connect(
+            sigc::mem_fun(this, &MRFDongle::handle_beep_done));
+        beep_transfer->submit();
+        std::cout << "Beep submitted" << std::endl;
+        pending_beep_length = 0;
+    }
+}
+
+uint8_t MRFDongle::alloc_message_id()
+{
+    if (free_message_ids.empty())
+    {
+        throw std::runtime_error("Out of reliable message IDs");
+    }
+    uint8_t id = free_message_ids.front();
+    free_message_ids.pop();
+    return id;
+}
+
+void MRFDongle::free_message_id(uint8_t id)
+{
+    free_message_ids.push(id);
+}
+
+void MRFDongle::handle_mdrs(AsyncOperation<void> &op)
+{
+    USB::BulkInTransfer &mdr_transfer = dynamic_cast<USB::BulkInTransfer &>(op);
+    mdr_transfer.result();
+    if ((mdr_transfer.size() % 2) != 0)
+    {
+        throw std::runtime_error("MDR transfer has odd size");
+    }
+    for (unsigned int i = 0; i < mdr_transfer.size(); i += 2)
+    {
+        signal_message_delivery_report.emit(mdr_transfer.data()[i],
+                                            mdr_transfer.data()[i + 1]);
+    }
+    mdr_transfer.submit();
+}
+
+// TODO see #222
+void MRFDongle::handle_message(AsyncOperation<void> &, USB::BulkInTransfer &transfer)
+{
+    transfer.result();
+    // if (transfer.size() > 2)
```

Comment:
Can we remove the commented out code?

Reply:
I would rather leave this here for reference when robot status handling gets tackled.

Reply:
Ah didn't realize that's what it's for. I'm fine leaving it then.